src/analyzer/analyzersilence.cpp

```diff
@@ -44,34 +51,51 @@ bool AnalyzerSilence::initialize(const AnalyzerTrack& track,
     }
 
     m_iFramesProcessed = 0;
-    m_bPrevSilence = true;
     m_iSignalStart = -1;
     m_iSignalEnd = -1;
 
     return true;
 }
 
-bool AnalyzerSilence::processSamples(const CSAMPLE* pIn, const int iLen) {
-    for (int i = 0; i < iLen; i += mixxx::kAnalysisChannels) {
-        // Compute max of channels in this sample frame
-        CSAMPLE fMax = CSAMPLE_ZERO;
-        for (SINT ch = 0; ch < mixxx::kAnalysisChannels; ++ch) {
-            CSAMPLE fAbs = fabs(pIn[i + ch]);
-            fMax = math_max(fMax, fAbs);
-        }
+// static
+SINT AnalyzerSilence::findFirstSoundInChunk(std::span<const CSAMPLE> samples) {
+    return std::distance(samples.begin(), first_sound(samples.begin(), samples.end()));
+}
 
-        bool bSilence = fMax < m_fThreshold;
+// static
+/// returns a std::reverse_iterator
+SINT AnalyzerSilence::findLastSoundInChunk(std::span<const CSAMPLE> samples) {
+    // -1 is required, because the distance from the fist sample index (0) to crend() is 1,
+    return std::distance(first_sound(samples.rbegin(), samples.rend()), samples.rend()) - 1;
+}
```

Comment:
I'm not too happy by the fact that `findLastSoundInChunk` returns `-1` when it found no while `findFirstSoundInChunk` returns `samples.end()` (which is the more common pattern for stl algorithms).

Reply:
This is due to the fact that it uses the revers iterators. `end()` points to the first out of bounds index `size()` and `rend()` points to the first out of bound index -1 when going reverse. So it kind of matches a common pattern.

Reply:
Right, but the the function doesn't return a reverse iterator but an index instead. So that information is effectively lost. I'd expect the function to behave like its counterpart. We might want to add a special case for the `first_sound(samples.rbegin(), samples.rend() == samples.rend()` case so we get consistent behavior across both functions.

Reply:
OK